src/main/groovy/com/morpheusdata/upcloud/sync/VirtualMachinesSync.groovy

```diff
@@ -0,0 +1,377 @@
+import com.morpheusdata.model.projection.ComputeServerIdentityProjection
+import com.morpheusdata.upcloud.UpcloudPlugin
+import com.morpheusdata.upcloud.services.UpcloudApiService
+import com.morpheusdata.core.util.ComputeUtility
+import com.morpheusdata.model.ComputeServerType
+import com.morpheusdata.model.StorageVolumeType
+import com.morpheusdata.core.util.SyncList
+import com.morpheusdata.upcloud.UpcloudProvisionProvider
+
+class VirtualMachinesSync {
+    private Cloud cloud
+    UpcloudPlugin plugin
+    private MorpheusContext morpheusContext
+
+    VirtualMachinesSync(Cloud cloud, UpcloudPlugin plugin) {
+        this.cloud = cloud
+        this.plugin = plugin
+        this.morpheusContext = plugin.morpheusContext
+    }
+
+    def execute() {
+        try {
+            def authConfig = plugin.getAuthConfig(cloud)
+            def apiResults = UpcloudApiService.listServers(authConfig)
+
+            if (apiResults.success == true) {
+                def servicePlans = morpheusContext.async.servicePlan.listIdentityProjections(
+                        new DataQuery().withFilter("provisionType", "upcloud")
+                        .withFilter("active", true)
+                )
+
+                def serverRecords = morpheusContext.async.computeServer.listIdentityProjections(
+                        new DataQuery().withFilter("account", cloud.account)
+                        .withFilter("zone", cloud)
+                )
+
+                SyncTask<ComputeServerIdentityProjection, Map, ComputeServer> syncTask = new SyncTask<>(serverRecords, apiResults.imageData as Collection<Map>) as SyncTask<ComputeServerIdentityProjection, Map, ComputeServer>
+                syncTask.addMatchFunction { ComputeServerIdentityProjection imageObject, Map cloudItem ->
+                    imageObject.externalId == cloudItem?.uuid
+                }.withLoadObjectDetailsFromFinder { List<SyncTask.UpdateItemDto<ComputeServerIdentityProjection, Map>> updateItems ->
+                    morpheusContext.async.computeServer.listById(updateItems.collect { it.existingItem.id } as List<Long>)
+                }.onAdd { itemsToAdd ->
+                    addMissingVirtualMachines(itemsToAdd)
+                }.onUpdate { List<SyncTask.UpdateItem<ComputeServer, Map>> updateItems ->
+                    updateMatchedVirtualMachines(updateItems)
+                }.onDelete { removeItems ->
+                    removeMissingVirtualMachines(removeItems)
+                }.start()
+            } else {
+                log.error "Error in getting images: ${apiResults}"
+            }
+        } catch(e) {
+            log.error("cachePublicTemplates error: ${e}", e)
+        }
+    }
+
+    private addMissingVirtualMachines(Collection<Map> addList) {
+        def authConfig = plugin.getAuthConfig(cloud)
+        def serverType = new ComputeServerType(code: 'upcloudUnmanaged')
+        def servicePlans = morpheusContext.async.servicePlan.listIdentityProjections(
```

Comment:
The service plans list has already been loaded in `execute()` function and should just be passed down to this function.